Добавлена лабораторная №2

Shardenn opened the pull request:
Сама hash функция простейшая, потому в одну ячейку массива записываются
разные слова (по заданию надо составить список из повторяющихся). В
составлении списка я проверяю linkedItems > 0, будто hash функция
составлена правильно и в одной ячейке списком тянутся одинаковые слова.
Нужно ли исправлять проверку под действительную, мою hash функцию и
разгребать список из НЕодинаковых слов, попавших в одну ячейку ввиду плохой has функции?

Shardenn committed:
Добавлена лабораторная №2

Сама hash функция простейшая, потому в одну ячейку массива записываются
разные слова (по заданию надо составить список из повторяющихся). В
составлении списка я проверяю linkedItems > 0, будто hash функция
составлена правильно и в одной ячейке списком тянутся одинаковые слова.
Нужно ли исправлять проверку под действительную, мою hash функцию и
разгребать список из НЕодинаковых слов?

tapoton commented:
Линейный список вам нужен как раз для того, чтобы разрешить коллизии. Коллизия – это ситуация, когда два **_разных_** ключа дали один и тот же хэш. Не нужно создавать линейный список из одинаковых элементов. Я хочу, чтобы ваша хэш-таблица для каждого слова хранила количество раз, которое вы пытались его туда записать.

tapoton commented on Vikarchuk/hash/hash/main.cpp:
```diff
@@ -0,0 +1,161 @@
+	static const int tableSize = 20;
+
+	item *hashTable[tableSize];
+	item *head = new item; // ��� ������ �� ����
```
Поскольку GitHub не работает с Windows-кодировками, то я, как вы можете, догадаться, не могу прочитать ваш комментарий.

Смею предположить, что это поле отвечает за результирующий список, который требуется в задаче.
Делать так не нужно: ваша хэш-таблица должна хранить лишь общие данные. Для возвращения списка используйте функцию, которая его создаст и вернет вам, а вы его храните, печатайте, проще говоря, делайте что хотите с ним.

tapoton commented on Vikarchuk/hash/hash/main.cpp:
```diff
@@ -0,0 +1,161 @@
+	{
+		for (int i = 0; i < tableSize; i++)
+		{
+			hashTable[i] = new item;
```
Создание элементов при инициализации объекта не требуется. Для каждого индекса, если для него еще не было вставлено значение, гораздо проще и рациональнее хранить `NULL`: таким образом вы не будете с ходу выделять ваши 20 ячеек памяти под `item`, а выделите только нужные и по мере заполнения таблицы. Более того, проверять указатель на пустоту гораздо проще, чем делать проверку `pointer -> word == " "`

tapoton commented on Vikarchuk/hash/hash/main.cpp:
```diff
@@ -0,0 +1,161 @@
+	struct item
+	{
+		string word;
+		int linkedItemsCount;
```
Как я уже написал в комментарии к pull-request'у, вместо непонятных переменных "размер линейного списка" в структуре, которая отвечает лишь за одно слово, за один элемент этого списка, храните переменную, которая указывает на то, сколько раз был вызван метод добавления этого слова.

Дело не только в целесообразности вашего алгоритма, но и в том, что класс/структура/объединение/etc должны представлять собой некоторый обособленный по смыслу и логически набор данных (и, возможно, действий). Всегда помните, что базовые принципы ООП были нагло "содраны" с реального мира. 

Эта же мысль относится и к комментарию ниже о `head`.

Shardenn committed:
Изменённая лабораторная №2

tapoton commented:
С коллизиями невнимательно работаете.
Введите в качестве входных данных 5 одинаковых слов и шестым -- слово из тех же букв, но в другом порядке. Результат выйдет пустым.

tapoton commented:
Следующая задача.
Бэктрекинг (поиск с возвратом):
Задан граф. Определить является ли связанным заданный граф.

Shardenn committed:
Добавлена лабораторная №3

Лабораторная №3 - определение связности графа. Использовался метод
поиска в ширину и пометки пройденных вершин.

tapoton commented:
С третьей все ок, но пока вторую не поправили, не засчитаю

tapoton commented:
Графы:
Требуется найти в связном графе остовное дерево минимально веса.
**Входные данные**
Первая строка входного файла содержит два натуральных числа n и m - количество вершин и ребер графа соответственно (1≤n≤20000, 0≤m≤100000). Следующие m строк содержат описание ребер по одному на строке. Ребро номер i описывается тремя натуральными числами bi, ei и wi - номера концов ребра и его вес соответственно (1≤bi,ei≤n, 0≤wi≤100000). Граф является связным.
**Выходные данные**
Выведите единственное целое число - вес минимального остовного дерева.

Shardenn committed:
Добавлена лабораторная №4

Нахождение минимального остова. Реализовано алгоритмом Прима.

tapoton commented on Vikarchuk/min_ostov/min_ostov/main.cpp:
```diff
@@ -0,0 +1,69 @@
+	visited_verticles.push_back(currnode);
+	visited_count++;
+
+	while (visited_count < n)
```
и не стыдно весь алгоритм в `main` реализовывать?

tapoton commented:
Внешние сортировки: В текстовом файле находятся m квадратных матриц. Создать выходной файл, в котором исходные матрицы расположены в порядке убывания произведений всех элементов каждой матрицы.

Shardenn committed:
Доделанная лабораторная №2

ПОправлено после того, как было показано вам.
